pkg/channeld/connection.go

```diff
@@ -423,28 +424,31 @@ func (c *Connection) receive() {
 		packetReceived.WithLabelValues(c.connectionType.String()).Inc()
 
 	*/
-
-	for bufPos := 0; bufPos < c.readPos; {
+	bufPos := 0
+	for bufPos = 0; bufPos < c.readPos; {
 		if c.readPacket(&bufPos) == nil {
-			return
+			break
 		}
 		if bufPos < c.readPos {
 			combinedPacketCount.WithLabelValues(c.connectionType.String()).Inc()
 		}
 	}
 
+	//move unhandled content to the front
+	copy(c.readBuffer, c.readBuffer[bufPos:c.readPos])
 	// Reset read position
-	c.readPos = 0
+	c.readPos = c.readPos - bufPos
 }
 
 func (c *Connection) readPacket(bufPos *int) *channeldpb.Packet {
 	tag := c.readBuffer[*bufPos : *bufPos+PacketHeaderSize]
 	if tag[0] != 67 {
 		c.readPos = 0
-		packetDropped.WithLabelValues(c.connectionType.String()).Inc()
-		c.Logger().Warn("invalid tag, the packet will be dropped",
+		connectionClosed.WithLabelValues(c.connectionType.String()).Inc()

```

Comment:
@laughxing could you please explain why we should close the connection when received invalid tag?

Reply:
> @laughxing could you please explain why we should close the connection when received invalid tag?

when there comes an invalid wire format on a **streaming** socket, we can not just throw away what we received so far and expect that will recover magicly,  in worst case ,this will cause the connection continue to serve without communicating any meaningful data for a long time. So I would like to close the connection to give a quick feedback.